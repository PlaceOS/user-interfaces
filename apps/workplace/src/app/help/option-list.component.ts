import { Component, OnInit } from '@angular/core';

import {
    ApplicationLink,
    BaseClass,
    HashMap,
    SettingsService,
} from '@placeos/common';

@Component({
    selector: 'a-help-option-list',
    template: `
        <div
            class="flex items-center flex-wrap h-full w-full bg-gray-200 bg-cover bg-center p-2 overflow-auto"
            [style.background-image]="
                'url(' + settings?.background + ')' | safe: 'style'
            "
        >
            <ng-container *ngFor="let item of tiles">
                <a
                    class="relative flex flex-1"
                    [href]="item.link || item.route"
                    [target]="item.link ? '_blank' : ''"
                    [style.min-width]="tile_width + '%'"
                    [style.height]="tile_height + '%'"
                >
                    <div
                        inner
                        class="relative rounded-lg flex sm:flex-col space-x-2 sm:space-x-0 items-center sm:justify-center bg-black/60 hover:bg-black/30 text-white overflow-hidden bg-center bg-cover m-auto shadow cursor-pointer"
                        [style.background-image]="
                            'url(' + item?.background + ')' | safe: 'style'
                        "
                    >
                        <div
                            overlay
                            class="absolute inset-0 bg-black/60 hover:bg-black/30 z-0"
                            *ngIf="item?.background"
                        ></div>
                        <app-icon
                            class="relative text-6xl sm:text-7xl z-10"
                            [icon]="item?.icon"
                        ></app-icon>
                        <div>
                            <p class="relative text-2xl z-10">
                                {{ item?.name }}
                            </p>
                            <p class="relative z-10">{{ item?.description }}</p>
                        </div>
                    </div>
                </a>
            </ng-container>
        </div>
    `,
    styles: [
        `
            [inner] {
                width: 786px;
                max-width: calc(100% - 1em);
                height: calc(100% - 1em);
                transition: box-shadow 200ms, background-color 200ms,
                    color 200ms;
            }

            [overlay] {
                transition: background-color 200ms;
            }

            .active [inner] {
                position: fixed;
                top: 0 !important;
                left: 0 !important;
                right: 0 !important;
                bottom: 0 !important;
                background-color: #fff;
                color: rgba(var(--primary), 0.85);
                z-index: 2;
            }
        `,
    ],
})
export class HelpOptionListComponent extends BaseClass {
    /** Settings for the help page */
    public settings: HashMap;
    /** Whether to show the overlay menu */
    public show_menu: boolean;
    /** Tiles to display on the help page */
    public get tiles(): ApplicationLink[] {
        return this._settings.get('app.help.tiles') || [];
    }

    /** Min width of a tile */
    public get tile_width(): number {
        const tiles = this.tiles;
        const columns = (this.settings || { columns: 1 }).columns || 1;
        if (tiles.length > 2) {
            return 100 / columns - 1;
        }
        return 99;
    }

    /** Height of each tile */
    public get tile_height(): number {
        const tiles = this.tiles;
        const columns = (this.settings || { columns: 1 }).columns || 1;
        if (tiles.length > 2) {
            return 100 / ((tiles.length + (tiles.length % columns)) / columns);
        }
        return 50;
    }

    constructor(private _settings: SettingsService) {
        super();
    }
}
