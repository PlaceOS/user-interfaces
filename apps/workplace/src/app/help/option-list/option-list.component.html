<div class="help-option-list" [style.background-image]="('url(' + settings?.background + ')') | safe:'style'">
    <div
        class="tile"
        *ngFor="let item of tiles"
        [style.min-width]="tile_width + '%'"
        [style.height]="tile_height + '%'"
    >
        <a *ngIf="item.link" class="link" [href]="item.link" target="_blank">
            <ng-container *ngTemplateOutlet="inner; context: { item: item }"></ng-container>
        </a>
        <a *ngIf="!item.link" class="route" [routerLink]="[item.route]" routerLinkActive="active">
            <ng-container *ngTemplateOutlet="inner; context: { item: item }"></ng-container>
        </a>
    </div>
</div>
<ng-template #inner let-item="item">
    <div class="inner" [style.background-image]="('url(' + item?.background + ')') | safe:'style'">
        <div class="overlay" *ngIf="item?.background"></div>
        <app-icon [icon]="item?.icon"></app-icon>
        <div class="text">{{ item?.name }}</div>
    </div>
</ng-template>
