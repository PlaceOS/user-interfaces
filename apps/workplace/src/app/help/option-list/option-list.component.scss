
@import 'variables';
@import 'mixins';

:host,
.help-option-list {
    display: flex;
    align-items: center;
    flex-wrap: wrap;
    width: 100%;
    height: 100%;
    background-color: $bg-light;
    background-position: center;
    background-size: cover;
    padding: 8px;
}

.tile {
    position: relative;
    flex: 1;
    border: 1px solid rgba(#fff, .2);

    &:hover {
        .inner,
        .overlay {
            background-color: rgba(#000, .35);
        }
    }

    @include respond-to(mobile) {
        font-size: .9em;
    }
}

app-icon {
    font-size: 5em;
    z-index: 1;
}

.inner {
    position: relative;
    width: 786px;
    max-width: calc(100% - 1em);
    height: calc(100% - 1em);
    margin: auto;
    border-radius: 8px;
    @include box-shadow;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    cursor: pointer;
    color: #fff;
    background-color: rgba(#000, .6);
    transition: box-shadow 200ms, background-color 200ms, color 200ms;
    background-position: center;
    background-size: cover;
    overflow: hidden;
}

.overlay {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: 0;
    background-color: rgba(#000, .6);
    transition: background-color 200ms;
}

.active {
    .inner {
        position: fixed;
        top: 0 !important;
        left: 0 !important;
        right: 0 !important;
        bottom: 0 !important;
        background-color: #fff;
        color: rgba($color-primary, .85);
        z-index: 2;
    }
}

a {
    text-decoration: none;
}

.text {
    font-size: 1.5em;
    z-index: 1;
}
