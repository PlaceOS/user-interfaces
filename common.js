"use strict";(self.webpackChunkworkplace=self.webpackChunkworkplace||[]).push([["common"],{7625:(W,T,t)=>{t.d(T,{I:()=>_});var L=t(3918),g=t(9638),D=t(8368),M=t(1299),I=t(6221),B=t(2886),U=t(3690),p=t(1810),O=t(9377),c=t(3200),u=t(4505),y=t(7716),C=t(823),o=t(8759),x=t(9151),P=t(6116),v=t(9095),l=t(3298),h=t(6942),a=t(9128),r=t(7418),K=t(9676);class _ extends D.KG{constructor(i,d){super(),this._settings=i,this._org=d,this._poll=new u.X(0),this._poll_type=new u.X("api"),this._loading=new u.X(!1),this._filters=new u.X({shown_types:["event","desk","parking","visitor"]}),this._date=new u.X(Date.now()),this._update=(0,y.aj)([this._date,this._poll]).pipe((0,C.b)(500),(0,o.b)(s=>this._loading.next(!0))),this._space_bookings=this._org.active_building.pipe((0,x.h)(s=>!!s),(0,P.g)("id"),(0,C.b)(300),(0,o.b)(s=>this.unsubWith("bind:")),(0,v.w)(({id:s})=>(this._loading.next(!0),(0,B.u2)(s))),(0,l.x)(([s],[e])=>s!==e),(0,v.w)(s=>(this._loading.next(!1),(0,y.aj)((s||[]).map(e=>{const n=(0,U.rTZ)(e.id,"Bookings").binding("bookings"),f=n.listen().pipe((0,h.U)(R=>(R||[]).map(A=>new M.ym({...A,resources:A.attendees.filter(m=>m.email===e.email||m.resource),system:e}))));return this.hasSubscription(`bind:${e.id}`)||this.subscription(`bind:${e.id}`,n.bind()),f})))),(0,h.U)(s=>(0,D.xH)(s)),(0,a.d)(1)),this.ws_events=this._space_bookings.pipe((0,h.U)(s=>{const e=(0,D.ar)();return s.filter(n=>n.host.toLowerCase()===e.email.toLowerCase()||n.attendees.find(f=>f.email.toLowerCase()===e.email.toLowerCase()))})),this.api_events=this._update.pipe((0,v.w)(([s])=>{const e={period_start:(0,p.Z)((0,O.Z)(s)),period_end:(0,p.Z)((0,c.Z)(s))};return this._settings.get("app.events.use_bookings")?(0,g.F2)({...e,type:"room"}).pipe((0,h.U)(n=>n.map(f=>(0,M.Yd)(f))),(0,r.K)(n=>[])):(0,M.u$)({...e}).pipe((0,r.K)(n=>[]))}),(0,o.b)(()=>this.timeout("end_loading",()=>this._loading.next(!1))),(0,a.d)(1)),this.events=this._poll_type.pipe((0,v.w)(s=>"api"===s?this.api_events:this.ws_events)),this.visitors=this._update.pipe((0,v.w)(([s])=>(0,g.F2)({period_start:(0,p.Z)((0,O.Z)(s)),period_end:(0,p.Z)((0,c.Z)(s)),type:"visitor"}).pipe((0,r.K)(e=>(console.error(e),[])))),(0,o.b)(()=>this.timeout("end_loading",()=>this._loading.next(!1))),(0,a.d)(1)),this.desks=this._update.pipe((0,v.w)(([s])=>(0,g.F2)({period_start:(0,p.Z)((0,O.Z)(s)),period_end:(0,p.Z)((0,c.Z)(s)),type:"desk"}).pipe((0,r.K)(e=>(console.error(e),[])))),(0,o.b)(()=>this.timeout("end_loading",()=>this._loading.next(!1))),(0,a.d)(1)),this.parking=this._update.pipe((0,v.w)(([s])=>(0,g.F2)({period_start:(0,p.Z)((0,O.Z)(s)),period_end:(0,p.Z)((0,c.Z)(s)),type:"parking"}).pipe((0,r.K)(e=>[]))),(0,o.b)(()=>this.timeout("end_loading",()=>this._loading.next(!1))),(0,a.d)(1)),this.bookings=(0,y.aj)([this.events,this.visitors,this.desks,this.parking]).pipe((0,h.U)(([s,e,n,f])=>[...s,...e,...n,...f].sort((R,A)=>R.date-A.date))),this.filtered_bookings=(0,y.aj)([this.bookings,this._filters]).pipe((0,h.U)(([s,e])=>s.filter(n=>n instanceof M.ym&&e?.shown_types?.includes("event")||e?.shown_types?.includes(n.booking_type)))),this.filters=this._filters.asObservable(),this.date=this._date.asObservable(),this.loading=this._loading.asObservable(),this.subscription("poll_type",this._org.active_building.subscribe(()=>this._poll_type.next(this._settings.get("app.schedule.use_websocket")?"ws":"api")))}triggerPoll(){this._poll.next(Date.now())}startPolling(i=6e4){return this.interval("poll",()=>{"visible"===document.visibilityState&&this._poll.next(Date.now())},i),()=>this.stopPolling()}stopPolling(){this.clearInterval("poll")}setDate(i){this._date.next(i)}removeItem(i){this._poll.next(Date.now())}toggleType(i,d=!1){var s=this;return(0,L.Z)(function*(){const e=s._filters.getValue()||{shown_types:[]},{shown_types:n}=e;n&&(n.includes(i)||d)?s._filters.next({...e,shown_types:n.filter(f=>f!==i)}):s._filters.next({...e,shown_types:[...n,i]})})()}}_.\u0275fac=function(i){return new(i||_)(K.LFG(D.gb),K.LFG(I.w7))},_.\u0275prov=K.Yz7({token:_,factory:_.\u0275fac,providedIn:"root"})},8676:(W,T,t)=>{t.d(T,{I:()=>r});var L=t(9638),g=t(9434),D=t(8368),M=t(1299),I=t(1810),B=t(9377),U=t(312),p=t(3200),O=t(4505),c=t(1942),u=t(7716),y=t(4350),C=t(9095),o=t(9128),x=t(823),P=t(522),v=t(6942),l=t(7418),h=t(8759),a=t(9676);class r extends D.KG{constructor(_){super(),this._settings=_,this._poll=new O.X(0),this._options=new O.X({start:Date.now()}),this._loading=new O.X(""),this._schedule=new O.X([]),this.options=this._options.asObservable(),this.loading=this._loading.asObservable(),this.schedule=this._loading.asObservable(),this.calendars=(0,c.H)(1e3).pipe((0,C.w)(E=>(0,g.yJ)()),(0,o.d)(1)),this.events=(0,u.aj)([this._options,this._poll]).pipe((0,x.b)(1e3),(0,P.zg)(([E])=>{this._loading.next("Loading schedule...");const i={period_start:(0,I.Z)((0,B.Z)(E.start)),period_end:(0,I.Z)((0,U.Z)((0,p.Z)(E.start),6))};return E.calendar&&(i.calendar=E.calendar),this._schedule.next(this._schedule.getValue().filter(d=>!(0,D.MZ)(1e3*i.period_start,1e3*i.period_end,d.date,d.date+60*d.duration*1e3))),(0,y.D)([!0===this._settings.get("app.events.use_bookings")?(0,L.F2)({...i,type:"room"}).pipe((0,v.U)(d=>d.map(s=>(0,M.Yd)(s)))):(0,M.u$)({...i}),(0,L.F2)({...i,type:"desk"}),(0,L.F2)({...i,type:"parking"})]).pipe((0,l.K)(d=>[]))}),(0,v.U)(([E,i])=>{const d=[...this._schedule.getValue(),...E,...i.filter(s=>"declined"!==s.status)].sort((s,e)=>s.date-e.date);return this._schedule.next((0,D.Tw)(d,"id")),d}),(0,l.K)(E=>[]),(0,h.b)(E=>this._loading.next("")),(0,o.d)(1))}startPolling(_=15e3){this.interval("poll",()=>this._poll.next(Date.now()),_)}stopPolling(){this.clearInterval("poll")}setOptions(_){this._options.next({...this._options.getValue(),..._})}}r.\u0275fac=function(_){return new(_||r)(a.LFG(D.gb))},r.\u0275prov=a.Yz7({token:r,factory:r.\u0275fac,providedIn:"root"})},9434:(W,T,t)=>{t.d(T,{ot:()=>P,yJ:()=>o.yJ}),t(4603);var g=t(3918),D=t(4505),M=t(8759),I=t(9128),B=t(5670),U=t(1810),p=t(9377),O=t(3200),c=t(2317),u=t(8939),y=t(719),C=t(565),o=t(6217),x=t(9676);class P extends u.K{constructor(l,h){super(),this._org=l,this._settings=h,this._calendars=new D.X([]),this.calendar_list=(0,o.yJ)().pipe((0,M.b)(a=>this._calendars.next(a)),(0,I.d)(1)),this.query=()=>(0,o.yJ)(),this.freeBusy=a=>(0,o.tS)(a,this._org),this.availability=a=>(0,o.v7)(a),this._org.initialised.pipe((0,B.P)(a=>a)).subscribe(()=>this.init())}init(){var l=this;return(0,g.Z)(function*(){l._settings.get("app.events.use_bookings")||l._initialised.next(!0)})()}get calendars(){return this._calendars.getValue()}getFreeBusyDate(l,h){return(0,o.tS)({period_start:(0,U.Z)((0,p.Z)(l)),period_end:(0,U.Z)((0,O.Z)(l)),calendars:h},this._org)}checkSpacesAvailability(l,h,a,r){return(0,g.Z)(function*(){const K=yield(0,o.v7)({period_start:h,period_end:a,system_ids:l.join(",")}).toPromise(),_=new Date(r?.date).valueOf(),E=(0,c.Z)(_,r?.duration).valueOf();return!!K.every(d=>{const s=d.availability;if(r&&d.id===r.system?.email){const e=s.findIndex(n=>n.date>=_&&(0,c.Z)(n.date,n.duration).valueOf()<=E);-1!==e&&s.splice(e,1)}return!s.length})})()}}P.\u0275fac=function(l){return new(l||P)(x.LFG(C.w),x.LFG(y.g))},P.\u0275prov=x.Yz7({token:P,factory:P.\u0275fac,providedIn:"root"})}}]);
//# sourceMappingURL=common.js.map