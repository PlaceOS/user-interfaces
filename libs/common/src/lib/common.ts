export { BaseClass } from './base.class';
export { GoogleAnalyticsService } from './google-analytics.service';
export { HotkeysService } from './hotkeys.service';
export { SettingsService } from './settings.service';
export { VorlonService } from './vorlon.service';
export { VERSION } from './version';

export * from './angular-animations';
export * from './api';
export * from './application';
export * from './formatting';
export * from './general';
export * from './notifications';
export * from './placeos';
export * from './room-configuration.interface';
export * from './settings';
export * from './types';
export * from './user-state';
export * from './spec-helpers';
