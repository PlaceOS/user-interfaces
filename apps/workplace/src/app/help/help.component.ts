import { Component, OnInit } from '@angular/core';

import { SettingsService } from '@placeos/common';

@Component({
    selector: 'app-help',
    template: `
        <topbar></topbar>
        <div class="flex-1 flex sm:flex-row flex-col-reverse h-1/2">
            <nav-menu class="relative z-10"></nav-menu>
            <main
                class="relative z-0 flex flex-col flex-1 h-1/2 sm:h-auto overflow-hidden"
            >
                <a-help-option-list class="h-full"></a-help-option-list>
            </main>
        </div>
    `,
    styles: [
        `
            :host {
                position: absolute;
                top: 0;
                left: 0;
                right: 0;
                bottom: 0;
                overflow: hidden;
                display: flex;
                flex-direction: column;
            }
        `,
    ],
})
export class HelpComponent implements OnInit {
    constructor(private _settings: SettingsService) {}

    public ngOnInit(): void {
        this._settings.title = 'Help';
    }
}
