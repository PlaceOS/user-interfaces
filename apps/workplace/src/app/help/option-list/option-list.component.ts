import { Component, OnInit } from '@angular/core';

import { ApplicationLink, BaseClass, HashMap, SettingsService } from '@user-interfaces/common';

@Component({
    selector: 'a-help-option-list',
    templateUrl: './option-list.component.html',
    styleUrls: ['./option-list.component.scss']
})
export class HelpOptionListComponent extends BaseClass implements OnInit {
    /** Settings for the help page */
    public settings: HashMap;
    /** Whether to show the overlay menu */
    public show_menu: boolean;
    /** Tiles to display on the help page */
    public get tiles(): ApplicationLink[] {
        return (this.settings || { tiles: [] }).tiles || [];
    }

    /** Min width of a tile */
    public get tile_width(): number {
        const tiles = this.tiles;
        const columns = (this.settings || { columns: 1 }).columns || 1;
        if (tiles.length > 2) {
            return 100 / columns - 1;
        }
        return 99;
    }

    /** Height of each tile */
    public get tile_height(): number {
        const tiles = this.tiles;
        const columns = (this.settings || { columns: 1 }).columns || 1;
        if (tiles.length > 2) {
            return 100 / ((tiles.length + tiles.length % columns) / columns );
        }
        return 50;
    }

    constructor(private _settings: SettingsService) {
        super();
    }

    public ngOnInit(): void {
        this.settings = this._settings.get('app.help') || {};
    }
}
