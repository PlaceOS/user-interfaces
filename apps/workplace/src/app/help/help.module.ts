import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule, Route } from '@angular/router';

import { HelpComponent } from './help.component';
import { HelpOptionListComponent } from './option-list.component';
import { SharedComponentModule } from '../components/shared.module';

const ROUTES: Route[] = [{ path: '', component: HelpComponent }];

@NgModule({
    declarations: [HelpComponent, HelpOptionListComponent],
    imports: [
        CommonModule,
        RouterModule.forChild(ROUTES),
        SharedComponentModule,
    ],
})
export class HelpModule {}
