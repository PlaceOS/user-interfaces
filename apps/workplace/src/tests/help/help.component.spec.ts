import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { SettingsService } from '@placeos/common';
import { MockComponent } from 'ng-mocks';
import { NavMenuComponent } from '../../app/components/nav-menu.component';
import { TopbarComponent } from '../../app/components/topbar.component';
import { HelpComponent } from '../../app/help/help.component';
import { HelpOptionListComponent } from '../../app/help/option-list.component';

describe('HelpComponent', () => {
    let spectator: Spectator<HelpComponent>;
    const createComponent = createComponentFactory({
        component: HelpComponent,
        providers: [{ provide: SettingsService, useValue: {} }],
        declarations: [
            MockComponent(TopbarComponent),
            MockComponent(NavMenuComponent),
            MockComponent(HelpOptionListComponent),
        ],
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(spectator.element).toMatchSnapshot();
    });
});
