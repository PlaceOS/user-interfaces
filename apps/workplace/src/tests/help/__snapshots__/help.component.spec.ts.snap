// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelpComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <topbar />
  <div
    class="flex-1 flex sm:flex-row flex-col-reverse h-1/2"
  >
    <nav-menu
      class="relative z-10"
    />
    <main
      class="relative z-0 flex flex-col flex-1 h-1/2 sm:h-auto overflow-hidden"
    >
      <a-help-option-list
        class="h-full"
      />
    </main>
  </div>
</div>
`;
